Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -324,61 +315,105 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
+			if (isPaletteTexture)
+				paletteDecodedID = TextureAddress::TMem(tmem_addr, tlutaddr);
+		}
+	}
+	else
+	{
+		texID = TextureAddress::Mem(address);
+		if (isPaletteTexture)
+			paletteDecodedID = TextureAddress::MemPalette(address, tlutaddr);
+	}
 
-		// 2. a) For EFB copies, only the hash and the texture address need to match
-		if (entry->IsEfbCopy() && tex_hash == entry->hash && address == entry->addr)
+	auto search_result = textures.find(texID);
+	bool palette_decoded_entry = false;
+	if (isPaletteTexture && search_result == textures.end())
+	{
+		search_result = textures.find(paletteDecodedID);
+		palette_decoded_entry = true;
+	}
+
+	TCacheEntryBase* entry = nullptr;
+	if (search_result != textures.end())
+		entry = search_result->second;
+
+	if (entry)
+	{
+		// EFB copies have slightly different rules: the hash doesn't need to match
+		// in EFB2Tex mode, and EFB copy formats have different meanings from texture
+		// formats.
+		if (entry->IsEfbCopy() && (g_ActiveConfig.bCopyEFBToTexture || tex_hash == entry->hash))
 		{
-			// TODO: Print a warning if the format changes! In this case,
-			// we could reinterpret the internal texture object data to the new pixel format
-			// (similar to what is already being done in Renderer::ReinterpretPixelFormat())
+			// TODO: We should check format/width/height/levels for EFB copies. Checking
+			// format is complicated because EFB copy formats don't exactly match
+			// texture formats. I'm not sure what effect checking width/height/levels
+			// would have.
+			if (!palette_decoded_entry && isPaletteTexture)
+			{
+				// Perform palette decoding.
+				// TODO: Skip decoding if we find a match.
+				auto decoded_entry_iter = textures.find(paletteDecodedID);
+				if (decoded_entry_iter != textures.end())
+				{
+					// Pool this texture and make a new one later.  We delay actually freeing
+					// the texture if it's currently bound; this can happen in edge cases
+					// involving multiple stages bound to the same address.
+					MaybeFreeTexture(decoded_entry_iter->second);
+					textures.erase(decoded_entry_iter);
+				}
+
+				TCacheEntryBase *decoded_entry = AllocateTexture(entry->config);
+
+				decoded_entry->SetGeneralParameters(paletteDecodedID, texture_size, full_format);
+				decoded_entry->SetDimensions(entry->native_width, entry->native_height, 1);
+				decoded_entry->SetHashes(TEXHASH_INVALID);
+				decoded_entry->frameCount = FRAMECOUNT_INVALID;
+
+				g_texture_cache->ConvertTexture(decoded_entry, entry, &texMem[tlutaddr], (TlutFormat)tlutfmt);
+				textures[paletteDecodedID] = decoded_entry;
+				entry = decoded_entry;
+			}
 			return ReturnEntry(stage, entry);
 		}
 
-		// 2. b) For normal textures, all texture parameters need to match
-		if (address == entry->addr && tex_hash == entry->hash && full_format == entry->format &&
+		// For normal textures, all texture parameters need to match.
+		if ((tex_hash ^ tlut_hash) == entry->hash && full_format == entry->format &&
```

Comment:
It would be still faster than to reupload the texture. But I also don't see the big usage. Maybe just upload all paletted as I8 first and convert them to C8 afterwards ;) So the I8 can be cached